RFC: build: support for pre-compiled headers.

theuni opened the pull request:
This cuts off roughly a third of my total build time, either with `-j8` or `-j24`.

There are many different ways to go about adding pre-compiled header support with CMake. This is probably the simplest and most naive, but it has a substantial impact.

I used [ClangBuildAnalyzer](https://github.com/aras-p/ClangBuildAnalyzer) to measure the headers which took the most amount of time to parse while building the kernel library. I did this under the assumption that those are probably the most included files in general, and because they should never be affected by changing defines/flags in different compilation units.

CMake re-generates a new pch file for each target that includes it, so it's only worth using for libraries/binaries which compile a substantial list of cpp files. I eyeballed the libs/bins which I thought made sense.

I chose to add `base_precompiled` as `PRIVATE` everywhere so that it's clear exactly where they're being used. 

Other approaches would involve keeping (or generating) lists of proper headers for each lib/bin, and potentially sharing them around when possible. The approach here (a monolithic interface lib) is unfortunately incompatible with sharing.

There's an additional (hackish) commit to support ccache + pch. Unfortunately this requires a rather new (4.8+) ccache in order to properly set the variables we need. With the required options turned on [as documented by ccache](https://ccache.dev/manual/latest.html#_precompiled_headers), this combo works fine for me locally.

Overall, I think this is a reasonable approach because it pretty much just works with no fiddling. The main drawback is that over time the headers list will grow stale, so it'll require some attention now and then.

---

Results: (`-DBUILD_UTIL_CHAINSTATE=ON -DBUILD_FUZZ_BINARY=ON -DBUILD_BENCH=ON`)

Before:
**** Time summary:
Compilation (1412 times):
  Parsing (frontend):         2619.3 s
  Codegen & opts (backend):   1553.3 s

Real make time:  6m20.894s

After:
**** Time summary:
Compilation (1422 times):
  Parsing (frontend):         1375.4 s
  Codegen & opts (backend):   1726.7 s

Real make time:  4m24.634s

pch + ccache 2nd run: 0m10.073s

theuni committed:
build: Add a rough attempt at pre-compiled headers

These are all kernel headers which required the most amount of frontend
parsing as measured by clang's -ftime-trace option.

theuni committed:
build: support ccache + pre-compiled headers when possible

ccache supports setting key=value options as of v4.8. Prior to that, setting
sloppiness would've required substantial hacking.

For 4.8+:
  - Set the required sloppiness options
  - Set "-Xclang -fno-pch-timestamp" for clang

Both options are documented here: https://ccache.dev/manual/latest.html

theuni committed:
build: display pch status in CMake output

theuni commented:
Ping @hebasto, @fanquake 

Also @willcl-ark, you might be interested in this as you've been benchmarking build times.

TheCharlatan commented:
Concept ACK

theuni committed:
guix: disable pre-compiled headers

fanquake commented:
MSAN failure was spurious, but not the tidy (https://cirrus-ci.com/task/4930868089716736):
```bash
[21:49:26.130] clang-tidy-18 -p=/ci_container_base/ci/scratch/build-x86_64-pc-linux-gnu -quiet -load=/tidy-build/libbitcoin-tidy.so /ci_container_base/src/validationinterface.cpp
[21:49:26.130] /ci_container_base/src/validationinterface.cpp:22:13: error: redundant 'RemovalReasonToString' declaration [readability-redundant-declaration,-warnings-as-errors]
[21:49:26.130]    22 | std::string RemovalReasonToString(const MemPoolRemovalReason& r) noexcept;
[21:49:26.130]       | ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[21:49:26.130] /ci_container_base/src/kernel/mempool_removal_reason.h:22:13: note: previously declared here
[21:49:26.130]    22 | std::string RemovalReasonToString(const MemPoolRemovalReason& r) noexcept;
[21:49:26.130]       |             ^
[21:49:26.130] ^^^ ⚠️ Failure generated from clang-tidy
[21:49:26.130] + echo '^^^ ⚠️ Failure generated from clang-tidy'
[21:49:26.130] + false
```

theuni commented:
Huh, I guess that happens because pch shuffles around the include order.

The redundant decl indeed seems broken to me. Will PR a quick fix.

theuni commented:
Combining with #30911 produces even more of a speedup (with Make, ninja is about the same).

Config: `-DCMAKE_BUILD_TYPE=Debug -DWITH_CCACHE=OFF -DBUILD_UTIL_CHAINSTATE=ON -DBUILD_FUZZ_BINARY=ON -DBUILD_BENCH=ON`
Using: `make -j24`

Master: 2m7.311s
This PR: 1m34.495s
This PR + #30911: 1m19.046s

Result: Completes in 62% of the time compared to master.

theuni committed:
ci: disable pch for tidy/iwyu builds

PCH is incompabible with iwyu:
  error: include-what-you-use does not support PCH

It seems likely that it'd mess with clang-tidy as well.

hebasto commented:
> Combining with #30911 produces even more of a speedup (with Make, ninja is about the same).

Why are ninja builds not affected?

theuni commented:
> > Combining with #30911 produces even more of a speedup (with Make, ninja is about the same).
> 
> Why are ninja builds not affected?

Ninja builds are sped up significantly by pch (this PR), it's the combo with #30911 doesn't seem to have much effect.

laanwj commented:
Concept ACK. It's nice that CMAKE proves a built-in abstraction for this so there's no need for really ugly stuff.

hebasto commented on CMakeLists.txt:
```diff
@@ -638,6 +638,12 @@ flags_summary()
+else()
+  set(pch_status ON)
+endif()
+message("Precompiled-headers ................... ${pch_status}")
```
nit: 
```suggestion
message("Precompiled headers ................... ${pch_status}")
```

hebasto reviewed (commented):
Concept ACK.

Tested and benchmarked on Ubuntu 23.10 with CMake 3.27.4 using the following command:
```
$ rm -rf build && cmake -B build -G Ninja --preset dev-mode -DWITH_MULTIPROCESS=OFF -DWITH_CCACHE=OFF && time cmake --build build -j 16
```

- The master branch @ 5fb94550638d0b01c184d2f3d5d97c8874c8c34b:
```
real	6m46.350s
```

- this PR @ 3cec3508a79c08db5b97e93bc4225ccdd7d32ad8:
```
real	6m10.958s
```

---

Do you want to keep the `base_precompiled` target or use the `PRECOMPILE_HEADERS` property on per-target basis?

edilmedeiros commented:
Concept ACK
Code LGTM.

---
`rm -rf build && cmake -B build -DBoost_INCLUDE_DIR=/opt/local/libexec/boost/1.78/include -DCMAKE_BUILD_TYPE=Debug -DWITH_CCACHE=OFF -DBUILD_BENCH=ON -DBUILD_FUZZ_BINARY=ON -DBUILD_UTIL_CHAINSTATE=ON && /usr/bin/time cmake --build build -j 11`

Master at ffe4261cb0669b1e1a926638e0498ae5b63f3599
1m56,95s real		16m57,22s user		1m31,97s sys

This PR at 3cec3508a79c08db5b97e93bc4225ccdd7d32ad8
1m32,76s real		12m59,11s user		1m6,45s sys

---
With ccache 4.10:

`rm -rf build && cmake -B build -DBoost_INCLUDE_DIR=/opt/local/libexec/boost/1.78/include -DCMAKE_BUILD_TYPE=Debug -DWITH_CCACHE=ON -DBUILD_BENCH=ON -DBUILD_FUZZ_BINARY=ON -DBUILD_UTIL_CHAINSTATE=ON && /usr/bin/time -h cmake --build build -j 11`

Master at ffe4261cb0669b1e1a926638e0498ae5b63f3599
2m18,47s real		19m11,32s user		2m39,58s sys

This PR at 3cec3508a79c08db5b97e93bc4225ccdd7d32ad8
1m52,09s real		15m5,11s user		2m14,64s sys

All running on a MacBook Pro M3 Pro.
---

Interesting that ccache made my build slower.

laanwj commented:
Complete build on RPi5 with NVME hat, 8GB memory:
```
$ cmake -B build  -G Ninja --preset=dev-mode  -DWITH_CCACHE=OFF --toolchain /data/tmp/bitcoin/depends/aarch64-unknown-linux-gnu/toolchain.cmake 
$ time ninja -C build
```

(PR with pch)
real    33m33.235s
user    127m24.712s
sys     4m22.753s

(base commit)
real    37m39.927s
user    144m17.888s
sys     3m26.133s

About 10% faster (in real time), not as impressive as some results on faster machines, but still nice to have.

fanquake commented on src/CMakeLists.txt:
```diff
@@ -75,6 +75,76 @@ add_subdirectory(secp256k1)
 set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 string(APPEND CMAKE_C_COMPILE_OBJECT " ${APPEND_CPPFLAGS} ${APPEND_CFLAGS}")
 
+
```
> Overall, I think this is a reasonable approach because it pretty much just works with no fiddling. The main drawback is that over time the headers list will grow stale, so it'll require some attention now and then.

Not sure if here, but it'd seem useful to document some (minified) version of the PR description in regards to maintaining this going forward. If anything to prevent if from being too arbtrary.

BrandonOdiwuor reviewed (commented):
Concept ACK using pre-compiled headers to speed up compilation

willcl-ark reviewed (commented):
Took a first pass over this. It seems to work for me locally, both with and without `ccache`.

I see both clang and gcc running faster with this change than merge-base @ 5fb94550638 (with clang being much faster than gcc in both cases):

```log
Summary
  env CCACHE_DISABLE=1 CC=clang CXX=clang++ just build (commit = 3cec3508a79c08db5b97e93bc4225ccdd7d32ad8) ran
    1.32 ± 0.02 times faster than env CCACHE_DISABLE=1 CC=clang CXX=clang++ just build (commit = 5fb94550638)
    1.43 ± 0.00 times faster than env CCACHE_DISABLE=1 CC=gcc CXX=g++ just build (commit = 3cec3508a79c08db5b97e93bc4225ccdd7d32ad8)
    1.53 ± 0.03 times faster than env CCACHE_DISABLE=1 CC=gcc CXX=g++ just build (commit = 5fb94550638)
```

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `env CCACHE_DISABLE=1 CC=clang CXX=clang++ just build (commit = 5fb94550638)` | 178.672 ± 2.623 | 176.287 | 181.481 | 1.32 ± 0.02 |
| `env CCACHE_DISABLE=1 CC=clang CXX=clang++ just build (commit = 3cec3508a79c08db5b97e93bc4225ccdd7d32ad8)` | 134.888 ± 0.242 | 134.672 | 135.150 | 1.00 |
| `env CCACHE_DISABLE=1 CC=gcc CXX=g++ just build (commit = 5fb94550638)` | 206.191 ± 4.083 | 203.754 | 210.905 | 1.53 ± 0.03 |
| `env CCACHE_DISABLE=1 CC=gcc CXX=g++ just build (commit = 3cec3508a79c08db5b97e93bc4225ccdd7d32ad8)` | 193.261 ± 0.185 | 193.056 | 193.416 | 1.43 ± 0.00 |


I too can't think of a cleaner way of implementing this, or making it more automated than using a manual list.

Left one comment about adding flags for gcc
